File: packages/sdkwork-utils-typescript/src/string.ts
const WORD_SPLIT = /[^a-zA-Z0-9]+/;
const CAMEL_BOUNDARY = /([a-z0-9])([A-Z])|([A-Z]+)([A-Z][a-z])/g;

function camelParts(input: string): string[] {
  const normalized = input.trim().replace(CAMEL_BOUNDARY, "$1$3 $2$4");
  return normalized
    .split(WORD_SPLIT)
    .filter(Boolean)
    .map((part) => part.toLowerCase());
}

export function isBlank(value: string | null | undefined): boolean {
  return value == null || value.trim().length === 0;
}

export function trim(value: string): string {
  return value.trim();
}

export function truncate(
  value: string,
  maxLen: number,
  suffix = "...",
): string {
  if (maxLen <= 0) {
    return "";
  }
  if (value.length <= maxLen) {
    return value;
  }
  if (suffix.length >= maxLen) {
    return suffix.slice(0, maxLen);
  }
  return `${value.slice(0, maxLen - suffix.length)}${suffix}`;
}

export function capitalize(value: string): string {
  if (value.length === 0) {
    return "";
  }
  return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
}

export function camelCase(value: string): string {
  const parts = camelParts(value);
  if (parts.length === 0) {
    return "";
  }
  return parts[0] + parts.slice(1).map(capitalize).join("");
}

export function snakeCase(value: string): string {
  return camelParts(value).join("_");
}

export function kebabCase(value: string): string {
  return camelParts(value).join("-");
}

export function slugify(value: string): string {
  return kebabCase(value)
    .replace(/[^a-z0-9-]/g, "")
    .replace(/^-+|-+$/g, "");
}

export function mask(
  value: string,
  visibleStart: number,
  visibleEnd: number,
  maskChar = "*",
): string {
  if (visibleStart + visibleEnd >= value.length) {
    return value;
  }
  return (
    value.slice(0, visibleStart) +
    maskChar.repeat(value.length - visibleStart - visibleEnd) +
    value.slice(value.length - visibleEnd)
  );
}

export function padStart(value: string, targetLen: number, padChar = " "): string {
  if (value.length >= targetLen) {
    return value;
  }
  return padChar.repeat(targetLen - value.length) + value;
}

export function padEnd(value: string, targetLen: number, padChar = " "): string {
  if (value.length >= targetLen) {
    return value;
  }
  return value + padChar.repeat(targetLen - value.length);
}

export function startsWith(value: string, prefix: string): boolean {
  return value.startsWith(prefix);
}

export function endsWith(value: string, suffix: string): boolean {
  return value.endsWith(suffix);
}

export function contains(value: string, substring: string): boolean {
  return value.includes(substring);
}

export function replaceAll(value: string, search: string, replacement: string): string {
  return value.split(search).join(replacement);
}

export function split(value: string, delimiter: string, trimParts = true): string[] {
  return value
    .split(delimiter)
    .map((part) => (trimParts ? part.trim() : part))
    .filter((part) => !trimParts || part.length > 0);
}

export function join(parts: string[], separator: string): string {
  return parts.join(separator);
}

export function repeat(value: string, count: number): string {
  if (count < 0) {
    throw new RangeError("repeat count must be >= 0");
  }
  return value.repeat(count);
}

export function normalizeWhitespace(value: string): string {
  return value.trim().split(/\s+/).filter(Boolean).join(" ");
}

const TEMPLATE_KEY = /\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g;

export function template(
  pattern: string,
  values: Record<string, string>,
): string {
  return pattern.replace(TEMPLATE_KEY, (match, key: string) => values[key] ?? match);
}

File: packages/sdkwork-utils-python/sdkwork_utils/string.py
import re

_WORD_SPLIT = re.compile(r"[^a-zA-Z0-9]+")
_CAMEL_BOUNDARY = re.compile(r"(?x)(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])")


def _camel_parts(value: str) -> list[str]:
    normalized = _CAMEL_BOUNDARY.sub(" ", value.strip())
    return [part.lower() for part in _WORD_SPLIT.split(normalized) if part]


def is_blank(value: str | None) -> bool:
    return value is None or value.strip() == ""


def trim(value: str) -> str:
    return value.strip()


def truncate(value: str, max_len: int, suffix: str = "...") -> str:
    if max_len <= 0:
        return ""
    if len(value) <= max_len:
        return value
    if len(suffix) >= max_len:
        return suffix[:max_len]
    return f"{value[: max_len - len(suffix)]}{suffix}"


def capitalize(value: str) -> str:
    if not value:
        return ""
    return value[0].upper() + value[1:].lower()


def camel_case(value: str) -> str:
    parts = _camel_parts(value)
    if not parts:
        return ""
    return parts[0] + "".join(capitalize(part) for part in parts[1:])


def snake_case(value: str) -> str:
    return "_".join(_camel_parts(value))


def kebab_case(value: str) -> str:
    return "-".join(_camel_parts(value))


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9-]", "", kebab_case(value))
    return slug.strip("-")


def mask(value: str, visible_start: int, visible_end: int, mask_char: str = "*") -> str:
    if visible_start + visible_end >= len(value):
        return value
    hidden = len(value) - visible_start - visible_end
    return f"{value[:visible_start]}{mask_char * hidden}{value[-visible_end:]}"


def pad_start(value: str, target_len: int, pad_char: str = " ") -> str:
    if len(value) >= target_len:
        return value
    return f"{pad_char * (target_len - len(value))}{value}"


def pad_end(value: str, target_len: int, pad_char: str = " ") -> str:
    if len(value) >= target_len:
        return value
    return f"{value}{pad_char * (target_len - len(value))}"


def starts_with(value: str, prefix: str) -> bool:
    return value.startswith(prefix)


def ends_with(value: str, suffix: str) -> bool:
    return value.endswith(suffix)


def contains(value: str, substring: str) -> bool:
    return substring in value


def replace_all(value: str, search: str, replacement: str) -> str:
    return value.replace(search, replacement)


def split(value: str, delimiter: str, trim_parts: bool = True) -> list[str]:
    parts = value.split(delimiter)
    if trim_parts:
        parts = [part.strip() for part in parts]
        parts = [part for part in parts if part]
    return parts


def join(parts: list[str], separator: str) -> str:
    return separator.join(parts)


def repeat(value: str, count: int) -> str:
    if count < 0:
        raise ValueError("repeat count must be >= 0")
    return value * count


def normalize_whitespace(value: str) -> str:
    return " ".join(value.split())


_TEMPLATE_KEY = re.compile(r"\{([a-zA-Z_][a-zA-Z0-9_]*)\}")


def template(pattern: str, values: dict[str, str]) -> str:
    return _TEMPLATE_KEY.sub(lambda match: values.get(match.group(1), match.group(0)), pattern)
